rotkehlchen/chain/evm/decoding/open_ocean/decoder.py

```diff
@@ -0,0 +1,188 @@
+import logging
+from abc import ABC
+from collections.abc import Callable
+from typing import TYPE_CHECKING, Any, Final
+
+from rotkehlchen.chain.ethereum.utils import (
+    token_normalized_value,
+    token_normalized_value_decimals,
+)
+from rotkehlchen.chain.evm.decoding.interfaces import DecoderInterface
+from rotkehlchen.chain.evm.decoding.structures import (
+    DEFAULT_DECODING_OUTPUT,
+    DecoderContext,
+    DecodingOutput,
+)
+from rotkehlchen.chain.evm.decoding.types import CounterpartyDetails
+from rotkehlchen.chain.evm.decoding.uniswap.constants import (
+    UNISWAP_V2_SWAP_SIGNATURE,
+    UNISWAP_V3_SWAP_SIGNATURE,
+)
+from rotkehlchen.chain.evm.decoding.utils import maybe_reshuffle_events
+from rotkehlchen.history.events.structures.types import HistoryEventSubType, HistoryEventType
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.types import ChecksumEvmAddress, EvmTransaction
+from rotkehlchen.utils.misc import bytes_to_address
+
+from .constants import CPT_OPENOCEAN, OPENOCEAN_EXCHANGE_ADDRESS, OPENOCEAN_LABEL
+
+if TYPE_CHECKING:
+    from rotkehlchen.assets.asset import Asset, CryptoAsset
+    from rotkehlchen.chain.evm.decoding.base import BaseDecoderTools
+    from rotkehlchen.chain.evm.node_inquirer import EvmNodeInquirer
+    from rotkehlchen.chain.evm.structures import EvmTxReceiptLog
+    from rotkehlchen.fval import FVal
+    from rotkehlchen.history.events.structures.evm_event import EvmEvent
+    from rotkehlchen.user_messages import MessagesAggregator
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+SWAPPED_TOPIC: Final = b'v\xaf"J\x148e\xa5\x0bAIn\x1asb&\x98i,V\\\x12\x14\xbc\x86/\x18\xe2-\x82\x9c^'  # noqa: E501
+
+
+class OpenOceanCommonDecoder(DecoderInterface, ABC):
+
+    def __init__(
+            self,
+            evm_inquirer: 'EvmNodeInquirer',
+            base_tools: 'BaseDecoderTools',
+            msg_aggregator: 'MessagesAggregator',
+    ) -> None:
+        super().__init__(
+            evm_inquirer=evm_inquirer,
+            base_tools=base_tools,
+            msg_aggregator=msg_aggregator,
+        )
+
+    def _get_asset_and_amount(
+            self,
+            asset_address: 'ChecksumEvmAddress',
+            raw_amount: int,
+    ) -> tuple['CryptoAsset', 'FVal']:
+        """Get asset and normalized amount from asset address and raw amount.
+        Specifically handles when asset is native rather than erc20 token.
+        """
+        asset = self.base.get_or_create_evm_asset(address=asset_address)
+        if asset == self.evm_inquirer.native_token:
+            amount = token_normalized_value_decimals(
+                token_amount=raw_amount,
+                token_decimals=18,
+            )
+        else:
+            amount = token_normalized_value(
+                token_amount=raw_amount,
+                token=asset.resolve_to_evm_token(),
```

Comment:
it is already an evm token right? No need to resolve? You can also do 

```
            amount = token_normalized_value_decimals(
                token_amount=raw_amount,
                token_decimals=18 if asset == self.evm_inquirer.native token else asset.decimals,
            )
```

Reply:
No its only a CryptoAsset - notice that I'm using `get_or_create_evm_asset` not `get_or_create_evm_token`, so that it doesn't throw a NotErc20Compliant error when the asset is native ETH. With that said, I can simplify it a bit and only use `token_normalized_value_decimals`, but I do still have to resolve to evm token to get decimals when its not the native asset.